src/test/kotlin/io/github/rybalkinsd/kohttp/dsl/UploadDslKtTest.kt

```diff
@@ -39,4 +54,47 @@ class UploadDslKtTest {
         assertEquals(200, uploadResponse.code())
         assertEquals(5243053, uploadResponse.body()?.string().asJson()["headers"]["content-length"].asInt())
     }
+
+    @Test
+    fun `test file upload with headers and parameters`() {
+
+        val expectedHeaders = mapOf(
+                "one" to "42",
+                "cookie" to "aaa=bbb; ccc=42"
+        )
+
+        val expectedArgs = mapOf(
+                "listOfParams" to listOf("val1", "val2", "val3"),
+                "query" to "cat"
+        )
+
+
+        val r = upload {
+
+            header {
+                "one" to 42
+                cookie {
+                    "aaa" to "bbb"
+                    "ccc" to 42
+                }
+            }
+
+            param {
+                "listOfParams" to listOf("val1", "val2", "val3")
+                "query" to "cat"
+            }
+
+            url("http://postman-echo.com/post")
+            val fileUri = this.javaClass.getResource("/cat.gif").toURI()
+            file(fileUri)
+        }
+
+        val parsedResponse = r.body()?.string().asJson()
+
+        assertEquals(expectedArgs["query"], parsedResponse["args"]["query"].asText())
+        assertEquals(expectedArgs["listOfParams"] as List<String>, parsedResponse["args"]["listOfParams"].asIterable().map { it.asText() })
+        assertResponses(expectedHeaders, parsedResponse["headers"])
+        assertEquals(1046214, parsedResponse["headers"]["content-length"].asInt())
+        assertTrue { parsedResponse["headers"]["content-type"].asText().startsWith("multipart/mixed; boundary=") }
```

Comment:
I'm thinking also about switching to junit5 in one of next releases...
It would be easier to do such checks (at least a bit easier syntax)

m/b good thing to investigate

Reply:
Agreed!